### PR TITLE
Minimal `Logs` integration

### PR DESCRIPTION
**Pull request by Halbaroth**
This PR adds:
- `Logs` as a new dependency;
- Logs' sources for each debug flag;
- A reporter that prints on the appropriate output (regular or
  diagnostic) according to the source.

As explained in #785, we cannot do really better without breaking the API of `Printer`

This PR is rebased on #1204 

See #1207 for an example using sources introduced here.

Notice that I create two sources for theories with a Shostak and Relation modules. An alternative implementation consists in prefixing all the messages in `*_rel.ml` by `[Rel]` as it does in the legacy code. I have no strong opinion on this choice.

I did not create a bijection between debug flags and sources because it is more flexible to use two different lists. 
For instance, it could be useful to create flags that set different sources (and actually, `--dd adt` for instance set `adt` and `adt_rel` sources).

### CODE REVIEW

File: src/bin/js/worker_interface.ml [hunk 1]
```diff
@@ -185,7 +185,7 @@ type options = {
   debug_cc : bool option;
   debug_combine : bool option;
   debug_constr : bool option;
-  debug_explanations : bool option;
+  debug_explanation : bool option;
```
Comment (bclement-ocp) [inline src/bin/js/worker_interface.ml]: The correct name for this feature is the plural "explanations"; please revert this change.

File: src/bin/common/parse_command.ml [hunk 2]
```diff
@@ -213,50 +213,107 @@ module Debug = struct
     | Commands -> "commands"
     | Optimize -> "optimize"
 
+  let set_level src = Logs.Src.set_level src (Some Debug)
+
   let mk ~verbosity flags =
+    let module S = Sources in
```
Comment (bclement-ocp) [inline src/bin/common/parse_command.ml]: I am not a big fan of using a global `Sources` module with all sources stuffed in. Since we are setting up sources in the `Parse_command` module which has everything has a dependency, would it be possible to have each module expose its own `val src : Logs.src` instead (like in the `Logs` examples in the documentation)?

Comment (Halbaroth) [reply src/bin/common/parse_command.ml]: Sure, this was my first design in fact...

Comment (Halbaroth) [reply src/bin/common/parse_command.ml]: Just, notice that some of them do not correspond with a module.

File: src/bin/js/worker_interface.ml (hunk 1 above)
Comment (Halbaroth) [reply src/bin/js/worker_interface.ml]: Please justify `correct`. The name of module is `explanation`. I did this modification because the current name of the debug flag is `explanation` too.

File: src/bin/common/parse_command.ml (hunk 2 above)
Comment (bclement-ocp) [reply src/bin/common/parse_command.ml]: I'd say that the few that do not correspond with a module can either be:
 - Stuffed into `options.ml` for now, or
 - Moved to a relevant module (e.g. we could have `fm_src` in `intervalCalculus.ml`), or
 - Removed entirely as they are not used (we will figure out what to do with them at a later time)

File: src/bin/js/worker_interface.ml (hunk 1 above)
Comment (bclement-ocp) [reply src/bin/js/worker_interface.ml]: > the current name of the debug flag is `explanation` too

I'd argue that this is incorrect as well (although I never noticed since I always write `-d ex`) ;)

What I mean is — there are multiple explanations used by Alt-Ergo that we want to debug; similar to `debug_commands`, `debug_warnings`, `debug_triggers`, `debug_types`, `debug_intervals`, etc. On the other hand the singular flags such as `debug_bitv`, `debug_interpretation`, `debug_matching`, `debug_uf`, etc. refer to a single data-structure / theory / step to debug (ignoring the fact that we duplicate the environment for case-split purposes).

"The correct name" was a bit strong, and I apologize — I don't think it matters too much either way. However I do believe that if it does not matter, it is best to not change whatever is there. Now that the change has been made let's keep it.

Comment (Halbaroth) [reply src/bin/js/worker_interface.ml]: I don't think it's matter too. When I see a discrepancy in our naming conventions and I am modifying this part of the code, I try to reduce it a bit. I agree to rename `explanation` to `explanations` everywhere (but for the module name of course).

Comment (Halbaroth) [reply src/bin/js/worker_interface.ml]: In fact, I removed the source from `Explanation` because `--d explanations` has no relation with this module. It increases the verbosity of explanations. We may add flags like `--verbosity explanations, ...`.

File: src/bin/common/parse_command.ml [hunk 3]
```diff
@@ -154,7 +154,8 @@ module Debug = struct
     | Cc
     | Combine
     | Constr
-    | Explanation
+    | Explanation (* deprecated *)
+    | Explanations
```
Comment (bclement-ocp) [inline src/bin/common/parse_command.ml]: We can simply use "explanations". Cmdliner uses prefix completion so that `-d explanation` will still work the same (on the other hand having both "explanation" and "explanations" breaks `-d ex`).

File: src/lib/reasoners/relation.ml [hunk 4]
```diff
@@ -41,6 +41,9 @@ module Rel5 : Sig_rel.RELATION = Adt_rel
 
 module Rel6 : Sig_rel.RELATION = Ite_rel
 
+(* This source is unused. *)
+let src = Logs.Src.create ~doc:__MODULE__ ("AltErgoLib." ^ __MODULE__)
```
Comment (bclement-ocp) [inline src/lib/reasoners/relation.ml]: Maybe we don't need it then? I think all the other sources are also unused, why single this one out?

Comment (Halbaroth) [reply src/lib/reasoners/relation.ml]: It is an artefact. I added it because I tried to add the source in the signature of `Sig_rel`, which was not a good idea. We do not want to build all the modules in `Parse_commands` just to setup up sources.

File: src/bin/common/parse_command.ml (hunk 3 above)
Comment (Halbaroth) [reply src/bin/common/parse_command.ml]: Oh I do not know we can do that :) thanks

File: src/lib/reasoners/relation.ml (hunk 4 above)
Comment (Halbaroth) [reply src/lib/reasoners/relation.ml]: In fact, it was not an artefact. I can remove `src` from `Sig_rel` but I have to introduce an extra signature in all the modules `*_rel.mli`.

**Review by bclement-ocp — APPROVED** [review]
